src/ORM/DataObject.php

```diff
@@ -742,9 +739,10 @@ public function setClassName($className)
      * there is no record, or the record has no ID. In this case, we do have an ID but
      * we still need to repopulate the defaults.
      *
-     * @param string $newClassName The name of the new class
+     * @template T of DataObject
```

Comment:
Does defining `@template` on `newClassInstance()` mean that only `T` is only used if the DataObject was instantiated using this method? i.e. there is no `@template` if I go `$obj = new MyDataObject();`?

Is this why methods like `DataObject::update()` return `static` instead of `T`? Or is that simply because `static` will always be correct so no need to be cute and `@return T`?

Reply:
I don't understand your question, but it seems to revolve around how the `@template T` used in this method's PHPDoc differs from the one in the class PHPDoc, so I'll try to explain that.

The class PHPDoc's `@template T` applies to the whole class, unless otherwise stated. Everywhere that has `T` as a type refers to the `T` in the class phpdoc unless there's a `@template T` in a narrower scope.

Using `@template T` here in the scope of this method's PHPDoc gives us a _new, separate_ `T` type. This `T` type will resolve to whatever is passed in to the `$newClassName` argument, since that has a typehint of `class-string<T>`

In other words, even if this is a `DataList<Member>`, when calling this method, the return type will _always_ resolve to whatever class is passed into this method. So you end up with the equivalent of this:

```php
/** @var DataList<Member> $list */
$list = Member::get();
/** @var Group $record */
$list->newClassInstance(Group::class);
```

If that hasn't answered your question, please try rewording it.

Reply:
> Is this why methods like `DataObject::update()` return `static` instead of `T`? Or is that simply because `static` will always be correct so no need to be cute and `@return T`?

`DataObject::update()` doesn't have any way to identify what `T` would be - so it doesn't make sense to try to use generics there. If the `update()` method took a classname variable and always returned the type of object that classname represents, we could use generics there - but since there's nothing like that, we just use `static`.